// sheweny/build/structures/Button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = void 0;
const collection_data_1 = require("collection-data");
/**
 * Represent a button
 * @class Button structure
 * @abstract
 */
class Button {
    /**
     * @constructor
     * @param {ShewenyClient} client - The client
     * @param {string[]} customId - The different buttons customid
     */
    constructor(client, customId) {
        this.client = client;
        this.customId = customId;
    }
    /**
     * Unregister a button
     * @public
     * @returns {boolean}
     */
    unregister() {
        this.client.buttons?.delete(this.customId);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a button
     * @public
     * @async
     * @returns {Promise<Collection<string[], Button> | null>} The buttons collection
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a button
     * @public
     * @async
     * @returns {Collection<string[], Button>} The buttons collection
     */
    async register() {
        const Button = (await Promise.resolve().then(() => require(this.path))).default;
        const btn = new Button(this.client);
        return this.client.buttons
            ? this.client.buttons.set(btn.customId, btn)
            : new collection_data_1.Collection().set(btn.customId, btn);
    }
}
exports.Button = Button;
