// sheweny/build/structures/SelectMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectMenu = void 0;
const collection_data_1 = require("collection-data");
/**
 * Represent a select menu
 * @class SelectMenu structure
 * @abstract
 */
class SelectMenu {
    /**
     * @constructor
     * @param {ShewenyClient} client - The client
     * @param {string[]} customId - The different select menu customid
     */
    constructor(client, customId) {
        this.client = client;
        this.customId = customId;
    }
    /**
     * Unregister a select menu
     * @public
     * @returns {boolean}
     */
    unregister() {
        this.client.selectMenus?.delete(this.customId);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a select menu
     * @public
     * @async
     * @returns {Promise<Collection<string[], SelectMenu> | null>} The select menus collection
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a select menu
     * @public
     * @async
     * @returns {Collection<string[], SelectMenu>} The select menus collection
     */
    async register() {
        const SelectMenu = (await Promise.resolve().then(() => require(this.path))).default;
        const sm = new SelectMenu(this.client);
        return this.client.selectMenus
            ? this.client.selectMenus.set(sm.customId, sm)
            : new collection_data_1.Collection().set(sm.customId, sm);
    }
}
exports.SelectMenu = SelectMenu;
